// steps_wizard/steps_wizard.js
const STATE = require('STATE')
const statedb = STATE(__filename)
const { get } = statedb(fallback_module)
const DOCS = require('DOCS')
const net = require('net_helper')

module.exports = steps_wizard

async function steps_wizard (opts, invite) {
  const { id, sdb } = await get(opts.sid)
  const { drive } = sdb

  const on = {
    style: inject
  }

  let currentActiveStep = 0
  const docs = DOCS(__filename)(opts.sid)
  const { io, _ } = net(id)
  docs.set_sys({ _, sdb, drive })

  io.on = {
    up: io_up()
  }
  if (invite) io.accept(invite)

  const el = document.createElement('div')
  const shadow = el.attachShadow({ mode: 'closed' })
  shadow.innerHTML = `
  <div class="steps-wizard main">
    <div class="steps-container">
      <div class="steps-slot"></div>
    </div>
  </div>`
  const sheet = new CSSStyleSheet()
  shadow.adoptedStyleSheets = [sheet]
  const steps_wizard_main = shadow.querySelector('.steps-wizard')
  const steps_entries = shadow.querySelector('.steps-slot')
  await sdb.watch(onbatch)

  return el

  function io_up () {
    return function onmessage ({ type, data }) {
      // docs_toggle handled globally by DOCS module
      if (type === 'init_data' && data) {
        render_steps(data, true)
      }
    }
  }

  function render_steps (steps, auto_focus_first) {
    if (!steps) { return }

    const is_single_step = steps.length === 1
    steps_wizard_main.style.display = is_single_step ? 'none' : ''

    steps_entries.innerHTML = ''
    currentActiveStep = 0

    steps.forEach(create_step_button)

    function create_step_button (step, index) {
      const btn = document.createElement('button')
      btn.className = 'step-button'
      btn.textContent = step.name + (step.type === 'optional' ? ' *' : '')
      btn.title = btn.textContent
      btn.setAttribute('data-step', index + 1)

      const accessible = can_access(index, steps)

      let status = 'default'
      if (!accessible) status = 'disabled'
      else if (step.is_completed) status = 'completed'
      else if (step.status === 'error') status = 'error'
      else if (step.type === 'optional') status = 'optional'

      btn.classList.add(`step-${status}`)

      if (index === currentActiveStep - 1 && index > 0) {
        btn.classList.add('back')
      }
      if (index === currentActiveStep + 1 && index < steps.length - 1) {
        btn.classList.add('next')
      }
      if (index === currentActiveStep) {
        btn.classList.add('active')
      }

      btn.__fn = on_step_click
      btn.onclick = docs.wrap_isolated(
        'function (event, sys) { event.currentTarget.__fn(event, sys) }',
        get_doc_content,
        { run_in_docs_mode: false }
      )

      async function on_step_click (event, sys) {
        currentActiveStep = index
        center_step(btn)
        render_steps(steps, false)
        sys._.up('step_clicked', {}, { ...step, index, total_steps: steps.length, is_accessible: accessible })
      }

      async function get_doc_content () {
        const doc_file = await drive.get('docs/README.md')
        return doc_file.raw || 'No documentation available'
      }

      steps_entries.appendChild(btn)

      if (auto_focus_first && index === 0) {
        btn.classList.add('active')
        center_step(btn)
        _.up('step_clicked', {}, { ...step, index: 0, total_steps: steps.length, is_accessible: accessible })
      }
    }
  }

  function center_step (step_button) {
    const container_width = steps_entries.clientWidth
    const step_left = step_button.offsetLeft
    const step_width = step_button.offsetWidth

    const center_position = step_left - (container_width / 2) + (step_width / 2)

    steps_entries.scrollTo({
      left: center_position,
      behavior: 'smooth'
    })
  }

  function can_access (index, steps) {
    for (let i = 0; i < index; i++) {
      if (!steps[i].is_completed && steps[i].type !== 'optional') {
        return false
      }
    }

    return true
  }

  async function onbatch (batch) {
    for (const { type, paths } of batch) {
      const data = await Promise.all(paths.map(load_path_raw))
      const func = on[type] || fail
      func(data, type)
    }

    function load_path_raw (path) { return drive.get(path).then(read_drive_file_raw) }
    function read_drive_file_raw (file) { return file.raw }
  }
  function fail (data, type) { console.warn('invalid message', { cause: { data, type } }) }

  function inject (data) { sheet.replaceSync(data[0]) }
}

function fallback_module () {
  return {
    api: fallback_instance,
    _: {
      DOCS: {
        $: ''
      },
      net_helper: {
        $: ''
      }
    }
  }

  function fallback_instance () {
    return {
      _: {
        DOCS: {
          0: ''
        },
        net_helper: {
          0: ''
        }
      },
      drive: {
        'docs/': {
          'README.md': {
            $ref: 'README.md'
          }
        },
        'style/': {
          'stepswizard.css': {
            $ref: 'stepswizard.css'
          }
        }
      }
    }
  }
}
