// theme_widget/theme_widget.js
const STATE = require('STATE')
const statedb = STATE(__filename)
const { get } = statedb(fallback_module)
const net = require('net_helper')

const program_container = require('program_container')
const taskbar = require('taskbar')

module.exports = theme_widget

async function theme_widget (opts, invite) {
  const { id, sdb } = await get(opts.sid)
  const { drive } = sdb

  const on = {
    style: inject,
    focused: handle_focused
  }

  // Inline focus tracking (merged from focus_tracker)
  let last_focused = null
  const { io, _ } = net(id)

  const el = document.createElement('div')
  const shadow = el.attachShadow({ mode: 'closed' })
  shadow.innerHTML = `
  <div class="theme-widget main">
    <div class="program-container-slot"></div>
    <div class="taskbar-slot"></div>
  </div>
  `
  const sheet = new CSSStyleSheet()
  shadow.adoptedStyleSheets = [sheet]
  const program_container_slot = shadow.querySelector('.program-container-slot')
  const taskbar_slot = shadow.querySelector('.taskbar-slot')

  const subs = await sdb.watch(onbatch)

  let program_container_el = null
  let taskbar_el = null
  io.on = {
    up: onmessage_from_root,
    program_container: program_container_protocol,
    taskbar: taskbar_protocol
  }
  if (invite) io.accept(invite)

  taskbar_el = await taskbar({ ...subs[1] }, io.invite('taskbar', { up: id }))
  taskbar_el.classList.add('taskbar')
  taskbar_slot.replaceWith(taskbar_el)

  program_container_el = await program_container({ ...subs[0] }, io.invite('program_container', { up: id }))
  program_container_el.classList.add('program-container')
  program_container_slot.replaceWith(program_container_el)

  return el

  async function onbatch (batch) {
    for (const { type, paths } of batch) {
      const data = await Promise.all(paths.map(load_path_raw))
      const func = on[type] || fail
      func({ data, type })
    }

    function load_path_raw (path) { return drive.get(path).then(read_drive_file_raw) }
    function read_drive_file_raw (file) { return file.raw }
  }

  function inject ({ data }) { sheet.replaceSync(data[0]) }

  // Inline focus tracker: reads persisted focused value to keep last_focused in sync
  function handle_focused ({ data }) {
    const focused = typeof data[0] === 'string' ? JSON.parse(data[0]) : data[0]
    last_focused = focused.value
  }

  function fail ({ data, type }) { console.warn('invalid message', { cause: { data, type } }) }

  function onmessage_from_root (msg) {
    const action_handlers = {
      update_actions_for_app: root_update_actions_for_app,
      update_quick_actions_for_app: root_forward_taskbar,
      update_steps_wizard_for_app: root_forward_taskbar,
      add_link_tab: root_forward_taskbar,
      remove_link_tab: root_forward_taskbar
    }
    const handler = action_handlers[msg.type] || fail
    handler(msg)

    function root_update_actions_for_app (msg) {
      if (_.program_container) _.program_container(msg.type, msg.data, msg.head ? { cause: msg.head } : {})
      else setTimeout(root_retry_send_program_container, 500, msg)
    }

    function root_retry_send_program_container (msg) { _.program_container(msg.type, msg.data, msg.head ? { cause: msg.head } : {}) }
    function root_forward_taskbar (msg) { _.taskbar(msg.type, msg.data, msg.head ? { cause: msg.head } : {}) }
  }

  // Inline focus tracker: handles ui_focus messages from children
  function handle_ui_focus (msg) {
    if (last_focused !== msg.data.type) {
      _.up('focused_app_changed', msg.data, {})
    }
    drive.put('focused/current.json', { value: msg.data.type })
  }

  // ---------
  // PROTOCOLS
  // ---------
  function program_container_protocol (msg) {
    const action_handlers = {
      ui_focus: program_container_forward_ui_focus,
      set_doc_display_handler: program_container_forward_up,
      action_auto_completed: program_container_forward_up,
      action_complete: program_container_forward_up
    }
    const handler = action_handlers[msg.type] || program_container_forward_taskbar
    handler(msg)

    function program_container_forward_ui_focus (msg) { handle_ui_focus(msg) }
    function program_container_forward_up (msg) { _.up(msg.type, msg.data, msg.head ? { cause: msg.head } : {}) }
    function program_container_forward_taskbar (msg) { _.taskbar(msg.type, msg.data, msg.head ? { cause: msg.head } : {}) }
  }

  function taskbar_protocol (msg) {
    const action_handlers = {
      ui_focus: taskbar_forward_ui_focus,
      docs_toggle: taskbar_docs_toggle,
      set_docs_mode: taskbar_forward_up,
      action_auto_completed: taskbar_forward_up,
      action_complete: taskbar_forward_up,
      link_tab_close_clicked: taskbar_forward_up,
      link_tab_clicked: taskbar_forward_up
    }
    const handler = action_handlers[msg.type] || taskbar_forward_program_container
    handler(msg)

    function taskbar_forward_ui_focus (msg) { handle_ui_focus(msg) }
    function taskbar_forward_up (msg) { _.up(msg.type, msg.data, msg.head ? { cause: msg.head } : {}) }
    function taskbar_forward_program_container (msg) { _.program_container(msg.type, msg.data, msg.head ? { cause: msg.head } : {}) }
    function taskbar_docs_toggle (msg) { _.program_container(msg.type, msg.data, msg.head ? { cause: msg.head } : {}) }
  }
}

function fallback_module () {
  return {
    api: fallback_instance,
    _: {
      program_container: {
        $: ''
      },
      taskbar: {
        $: ''
      },
      net_helper: {
        $: ''
      }
    },
    drive: {}
  }

  function fallback_instance () {
    return {
      _: {
        program_container: {
          0: '',
          mapping: {
            style: 'style',
            icons: 'icons',
            commands: 'commands',
            scroll: 'scroll',
            actions: 'actions',
            hardcons: 'hardcons',
            files: 'files',
            highlight: 'highlight',
            active_tab: 'active_tab',
            entries: 'entries',
            runtime: 'runtime',
            mode: 'mode',
            flags: 'flags',
            keybinds: 'keybinds',
            undo: 'undo',
            focused: 'focused',
            temp_actions: 'temp_actions',
            temp_quick_actions: 'temp_quick_actions',
            prefs: 'prefs',
            variables: 'variables',
            data: 'data',
            docs: 'docs',
            docs_style: 'docs_style'
          }
        },
        taskbar: {
          0: '',
          mapping: {
            style: 'style',
            icons: 'icons',
            actions: 'actions',
            prefs: 'prefs',
            variables: 'variables',
            data: 'data',
            hardcons: 'hardcons',
            docs: 'docs'
          }
        },
        net_helper: {
          0: ''
        }
      },
      drive: {
        'style/': {
          'theme.css': {
            raw: `
              .theme-widget {
                display: flex;
                flex-direction: column;
                width: 100%;
                height: 100%;
                background: #131315;
                min-height: 0;
                min-width: 400px;
              }
              .program-container {
                flex: 1 1 auto;
                min-height: 0;
                height: 100%;
              }
              .taskbar {
                flex: 0 0 auto;
                width: 100%;
                z-index: 10;
              }
            `
          }
        },
        'flags/': {},
        'commands/': {},
        'icons/': {},
        'scroll/': {},
        'actions/': {},
        'hardcons/': {},
        'files/': {},
        'highlight/': {},
        'active_tab/': {},
        'entries/': {},
        'runtime/': {},
        'mode/': {},
        'keybinds/': {},
        'undo/': {},
        'focused/': {
          'current.json': {
            raw: { value: 'default' }
          }
        },
        'temp_actions/': {},
        'temp_quick_actions/': {},
        'prefs/': {},
        'variables/': {},
        'data/': {},
        'docs_style/': {},
        'docs/': {
          'README.md': {
            $ref: 'README.md'
          }
        }
      }
    }
  }
}
