// action_bar/action_bar.js
const STATE = require('STATE')
const statedb = STATE(__filename)
const { get } = statedb(fallback_module)
const DOCS = require('DOCS')
const net = require('net_helper')

const quick_actions = require('quick_actions')

module.exports = action_bar

async function action_bar (opts, invite) {
  const { id, sdb } = await get(opts.sid)
  const { drive } = sdb

  const on = {
    style: inject,
    icons: iconject
  }

  const el = document.createElement('div')
  const shadow = el.attachShadow({ mode: 'closed' })

  shadow.innerHTML = `
  <div class="container">
    <div class="action-bar-container main">
      <div class="command-history">
        <button class="icon-btn"></button>
      </div>
      <div class="quick-actions">
        <quick-actions></quick-actions>
      </div>
    </div>
  </div>`
  const sheet = new CSSStyleSheet()
  shadow.adoptedStyleSheets = [sheet]
  const history_icon = shadow.querySelector('.icon-btn')
  const quick_placeholder = shadow.querySelector('quick-actions')

  const { io, _ } = net(id)
  let console_icon = {}
  const docs = DOCS(__filename)(opts.sid)
  const subs = await sdb.watch(onbatch)

  let selected_action = null

  io.on = {
    up: onmessage,
    quick_actions: quick_actions_protocol
  }
  if (invite) io.accept(invite)

  history_icon.innerHTML = console_icon
  history_icon.onclick = docs.wrap(onhistory, get_doc_content)
  const element = await quick_actions({ ...subs[0] }, io.invite('quick_actions', { up: id }))
  quick_placeholder.replaceWith(element)

  const parent_handler = {
    load_actions,
    selected_action: parent_selected_action,
    show_submit_btn,
    hide_submit_btn,
    step_clicked: parent_step_clicked,
    update_quick_actions_for_app,
    update_quick_actions_input,
    action_submitted: parent__action_submitted,
    clean_up: parent__clean_up
  }

  return el

  async function get_doc_content () {
    const doc_file = await drive.get('docs/README.md')
    return doc_file.raw || 'No documentation available'
  }

  async function onbatch (batch) {
    for (const { type, paths } of batch) {
      const data = await Promise.all(paths.map(load_path_raw))
      const func = on[type] || fail
      func({ data, type })
    }

    function load_path_raw (path) { return drive.get(path).then(read_drive_file_raw) }
    function read_drive_file_raw (file) { return file.raw }
  }

  function fail ({ data, type }) { console.warn('Unknown message type:', type, data) }
  function inject ({ data }) { sheet.replaceSync(data[0]) }
  function iconject ({ data }) { console_icon = data[0] }

  async function onhistory () {
    _.up('console_history_toggle', null, {})
    // const head2 = [by, to, mid++]
    // _.up({ head: head2, refs, type: 'ui_focus', data: 'command_history' })
  }

  // -------------------------------
  // Protocol: quick actions
  // -------------------------------

  function quick_actions_protocol (msg) {
    const quick_handlers = {
      display_actions: quick_actions_display_actions,
      action_submitted: quick_actions_action_submitted,
      filter_actions: quick_actions_filter_actions,
      update_quick_actions_input,
      activate_steps_wizard: quick_actions_activate_steps_wizard,
      ui_focus_docs
    }

    const { type } = msg
    const handler = quick_handlers[type] || fail
    handler(msg)
  }

  function quick_actions_filter_actions (msg) { _.up(msg.type, msg.data, msg.head ? { cause: msg.head } : {}) }
  function quick_actions_display_actions (msg) {
    const { data } = msg
    _.up(msg.type, msg.data, msg.head ? { cause: msg.head } : {})
    const display = typeof data === 'string' ? data : data.display
    const reason = typeof data === 'string' ? '' : data.reason
    const should_clean = display === 'none' && reason !== 'selected'
    if (should_clean) {
      _.up('clean_up', selected_action, msg.head ? { cause: msg.head } : {})
    }
  }

  function quick_actions_action_submitted (msg) {
    _.quick_actions('deactivate_input_field', { reason: 'completed' }, msg.head ? { cause: msg.head } : {})
    _.up('action_submitted', { selected_action }, msg.head ? { cause: msg.head } : {})
  }

  function onmessage (msg) {
    const { type } = msg
    if (type === 'docs_toggle') {
      _.quick_actions(type, msg.data, msg.head ? { cause: msg.head } : {})
    } else {
      const handler = parent_handler[type] || fail
      handler(msg)
    }
  }

  function load_actions (msg) {
    // const { data } = msg
  }
  function parent_selected_action (msg) {
    _.quick_actions(msg.type, msg.data, msg.head ? { cause: msg.head } : {})
  }
  function show_submit_btn (msg) { _.quick_actions('show_submit_btn', null, msg.head ? { cause: msg.head } : {}) }
  function hide_submit_btn (msg) { _.quick_actions('hide_submit_btn', null, msg.head ? { cause: msg.head } : {}) }

  function update_quick_actions_for_app (msg) {
    const { data, type } = msg
    _.quick_actions(type, data, msg.head ? { cause: msg.head } : {})
  }

  function update_quick_actions_input (msg) {
    const { data } = msg
    selected_action = data || null
    _.quick_actions('update_input_command', data, msg.head ? { cause: msg.head } : {})
  }

  function quick_actions_activate_steps_wizard (msg) { _.up('activate_steps_wizard', msg.data, msg.head ? { cause: msg.head } : {}) }

  function parent_step_clicked (msg) {
    const { data } = msg
    _.quick_actions('update_current_step', data, msg.head ? { cause: msg.head } : {})
    _.up('render_form', data, msg.head ? { cause: msg.head } : {})
  }

  function parent__action_submitted (msg) {
    _.quick_actions('deactivate_input_field', { reason: 'completed' }, msg.head ? { cause: msg.head } : {})
    _.up('action_submitted', msg.data, msg.head ? { cause: msg.head } : {})
  }

  function parent__clean_up (msg) {
    _.quick_actions('deactivate_input_field', { reason: 'cancel' }, msg.head ? { cause: msg.head } : {})
    _.up('clean_up', msg.data, msg.head ? { cause: msg.head } : {})
  }

  function ui_focus_docs (msg) { _.up(msg.type, msg.data, msg.head ? { cause: msg.head } : {}) }
}

function fallback_module () {
  return {
    api: fallback_instance,
    _: {
      quick_actions: { $: '' },
      DOCS: { $: '' },
      net_helper: { $: '' }
    }
  }
  function fallback_instance () {
    return {
      _: {
        quick_actions: {
          0: '',
          mapping: {
            style: 'style',
            icons: 'icons',
            actions: 'actions',
            hardcons: 'hardcons',
            prefs: 'prefs',
            docs: 'docs'
          }
        },
        DOCS: {
          0: ''
        },
        net_helper: {
          0: ''
        }
      },
      drive: {
        'icons/': {
          'console.svg': {
            $ref: 'console.svg'
          }
        },
        'docs/': {
          'README.md': {
            $ref: 'README.md'
          }
        },
        'style/': {
          'theme.css': {
            raw: `
              .container {
                display: flex;
                flex-direction: column;
                width: 100%;
              }
              .action-bar-container {
                display: flex;
                flex-direction: row;
                flex-wrap: nowrap;
                align-items: center;
                background: #131315;
                padding: 8px;
                gap: 12px;
              }
              .command-history {
                display: flex;
                align-items: center;
              }
              .quick-actions {
                display: flex;
                flex: auto;
                flex-direction: row;
                flex-wrap: nowrap;
                align-items: center;
                min-width: 340px;
              }
              .hide {
                display: none;
              }
              
              .icon-btn {
                display: flex;
                min-width: 32px;
                height: 32px;
                border: none;
                background: transparent;
                cursor: pointer;
                flex-direction: row;
                justify-content: center;
                align-items: center;
                padding: 6px;
                border-radius: 6px;
                color: #a6a6a6;
              }
              .icon-btn:hover {
                background: rgba(255, 255, 255, 0.1);
              }
              svg {
                width: 20px;
                height: 20px;
              }
            `
          }
        },
        'actions/': {},
        'hardcons/': {},
        'prefs/': {},
        'variables/': {}
      }
    }
  }
}
