// tabsbar/tabsbar.js
const STATE = require('STATE')
const state_db = STATE(__filename)
const { get } = state_db(fallback_module)
const DOCS = require('DOCS')
const net = require('net_helper')

const tabs_component = require('tabs')
const task_manager = require('task_manager')

module.exports = tabsbar

async function tabsbar (opts, invite) {
  const { id, sdb } = await get(opts.sid)
  const { drive } = sdb

  const on = {
    style: inject,
    icons: inject_icons
  }

  let dricons = {}
  let docs_toggle_active = false
  const on_message = {
    docs_toggle: handle_docs_toggle,
    add_link_tab: handle_forward_tabs,
    remove_link_tab: handle_forward_tabs
  }
  const { io, _ } = net(id)
  const el = document.createElement('div')
  const shadow = el.attachShadow({ mode: 'closed' })
  const docs = DOCS(__filename)(opts.sid)
  // Register actions with DOCS system
  const actions_file = await drive.get('actions/command.json')
  if (actions_file.raw) {
    const actions_data = typeof actions_file.raw === 'string' ? JSON.parse(actions_file.raw) : actions_file.raw
    docs.register_actions(actions_data)
  }

  io.on = {
    up: onmessage,
    tabs: tabs_protocol,
    task_manager: task_manager_protocol
  }
  if (invite) {
    io.accept(invite)
    const data = {
      type: 'wizard_hat',
      sid: opts.sid
    }
    _.up('ui_focus', data, {})
  }

  shadow.innerHTML = `
  <div class="tabs-bar-container main">
  <button class="hat-btn"></button>
  <tabs></tabs>
  <task-manager></task-manager>
  <button class="bar-btn"></button>
  </div>`
  const sheet = new CSSStyleSheet()
  shadow.adoptedStyleSheets = [sheet]
  const hat_btn = shadow.querySelector('.hat-btn')
  const bar_btn = shadow.querySelector('.bar-btn')

  const subs = await sdb.watch(onbatch)

  function onload (svg) {
    const parser = new DOMParser()
    const doc = parser.parseFromString(svg, 'image/svg+xml')
    const svgElem = doc.documentElement
    hat_btn.replaceChildren(svgElem)
    hat_btn.onclick = docs.wrap(hat_click, get_doc_content)
  }
  if (dricons[0]) {
    onload(dricons[0])
  }
  if (dricons[2]) {
    const parser = new DOMParser()
    const doc = parser.parseFromString(dricons[2], 'image/svg+xml')
    const svgElem = doc.documentElement
    bar_btn.replaceChildren(svgElem)
    bar_btn.onclick = on_bar_btn_click

    function on_bar_btn_click () {
      docs_toggle_active = !docs_toggle_active
      // Send message to root module to set docs mode
      _.up('set_docs_mode', { active: docs_toggle_active }, {})
      // Also send docs_toggle notification for UI updates
      _.up('docs_toggle', { active: docs_toggle_active }, {})
      bar_btn.classList.toggle('active', docs_toggle_active)
      _.task_manager('docs_toggle', { active: docs_toggle_active }, {})
    }
  }
  const tabs = await tabs_component({ ...subs[0] }, io.invite('tabs', { up: id }))
  tabs.classList.add('tabs-bar')
  shadow.querySelector('tabs').replaceWith(tabs)

  const task_mgr = await task_manager({ ...subs[1] }, io.invite('task_manager', { up: id }))
  task_mgr.classList.add('bar-btn')
  shadow.querySelector('task-manager').replaceWith(task_mgr)

  return el

  async function get_doc_content () {
    const doc_file = await drive.get('docs/README.md')
    return doc_file.raw || 'No documentation available'
  }

  async function hat_click () {
    const data = {
      type: 'wizard_hat',
      sid: opts.sid
    }
    _.up('ui_focus', data, {})
  }
  function onmessage (msg) {
    const handler = on_message[msg.type] || onmessage_fail
    handler(msg)
  }

  function handle_docs_toggle (msg) { _.tabs(msg.type, msg.data, msg.head ? { cause: msg.head } : {}) }
  function handle_forward_tabs (msg) { _.tabs(msg.type, msg.data, msg.head ? { cause: msg.head } : {}) }

  function onmessage_fail () {
    // Handle other message types
  }

  function tabs_protocol (msg) { _.up(msg.type, msg.data, msg.head ? { cause: msg.head } : {}) }

  function task_manager_protocol (msg) { _.up(msg.type, msg.data, msg.head ? { cause: msg.head } : {}) }

  async function onbatch (batch) {
    for (const { type, paths } of batch) {
      const data = await Promise.all(paths.map(load_path_raw))
      const func = on[type] || fail
      func(data, type)
    }

    function load_path_raw (path) { return drive.get(path).then(read_drive_file_raw) }
    function read_drive_file_raw (file) { return file.raw }
  }

  function fail (data, type) { console.warn('invalid message', { cause: { data, type } }) }
  function inject (data) { sheet.replaceSync(data[0]) }
  function inject_icons (data) { dricons = data }
}

function fallback_module () {
  return {
    api: fallback_instance,
    _: {
      tabs: {
        $: ''
      },
      task_manager: {
        $: ''
      },
      DOCS: {
        $: ''
      },
      net_helper: {
        $: ''
      }
    }
  }

  function fallback_instance () {
    return {
      _: {
        tabs: {
          0: '',
          mapping: {
            icons: 'icons',
            variables: 'variables',
            scroll: 'scroll',
            style: 'style',
            docs: 'docs',
            actions: 'actions'
          }
        },
        task_manager: {
          0: '',
          mapping: {
            count: 'count',
            style: 'style',
            docs: 'docs',
            actions: 'actions'
          }
        },
        DOCS: {
          0: ''
        },
        net_helper: {
          0: ''
        }
      },
      drive: {
        'style/': {
          'theme.css': {
            raw: `
              .tabs-bar-container {
                display: flex;
                flex: inherit;
                flex-direction: row;
                flex-wrap: nowrap;
                align-items: stretch;
              }
              .tabs-bar {
                display: flex;
                flex: auto;
                flex-direction: row;
                flex-wrap: nowrap;
                align-items: stretch;
                width: 256px;
              }
              .hat-btn, .bar-btn {
                display: flex;
                min-width: 32px;
                border: none;
                background: #131315;
                cursor: pointer;
                flex-direction: row;
                justify-content: center;
                align-items: center;
              }
              .bar-btn.active {
                background: #2d4a6d;
              }
            `
          }
        },
        'icons/': {
          '1.svg': {
            $ref: 'hat.svg'
          },
          '2.svg': {
            $ref: 'hat.svg'
          },
          '3.svg': {
            $ref: 'docs.svg'
          }
        },
        'actions/': {
          'command.json': {
            raw: JSON.stringify([
              {
                name: 'New File',
                icon: 'file',
                status: {
                  pinned: true,
                  default: true
                },
                steps: [
                  {
                    name: 'Enter File Name',
                    type: 'mandatory',
                    is_completed: false,
                    component: 'input_test',
                    status: 'default',
                    data: '',
                    commands: [
                      { type: 'set_mode', data: { mode: 'search' } },
                      { type: 'set_search_query', data: { query: 'file' } }
                    ]
                  },
                  { name: 'Choose Location', type: 'mandatory', is_completed: false, component: 'form_input', status: 'default', data: '' }
                ]
              },
              {
                name: 'Open File',
                icon: 'folder',
                status: {
                  pinned: false,
                  default: true
                },
                steps: [
                  {
                    name: 'Select File',
                    type: 'mandatory',
                    is_completed: false,
                    component: 'form_input',
                    status: 'default',
                    data: '',
                    commands: [
                      { type: 'set_mode', data: { mode: 'default' } },
                      { type: 'clear_selection', data: {} }
                    ]
                  }
                ]
              },
              {
                name: 'Save File',
                icon: 'save',
                status: {
                  pinned: true,
                  default: false
                },
                steps: [
                  { name: 'Choose Location', type: 'mandatory', is_completed: false, component: 'form_input', status: 'default', data: '' },
                  { name: 'Enter File Name', type: 'mandatory', is_completed: false, component: 'form_input', status: 'default', data: '' }
                ]
              },
              {
                name: 'Settings',
                icon: 'gear',
                status: {
                  pinned: false,
                  default: true
                },
                steps: [
                  {
                    name: 'Configure Settings',
                    type: 'optional',
                    is_completed: false,
                    component: 'input_test',
                    status: 'default',
                    data: '',
                    commands: [
                      { type: 'set_flag', data: { flag_type: 'hubs', value: 'true' } }
                    ]
                  }
                ]
              },
              {
                name: 'Help',
                icon: 'help',
                status: {
                  pinned: false,
                  default: false
                },
                steps: [
                  { name: 'View Documentation', type: 'optional', is_completed: false, component: 'form_input', status: 'default', data: '' }
                ]
              },
              {
                name: 'Terminal',
                icon: 'terminal',
                status: {
                  pinned: true,
                  default: true
                },
                steps: [
                  { name: 'Open Terminal', type: 'mandatory', is_completed: false, component: 'form_input', status: 'default', data: '' }
                ]
              },
              {
                name: 'Search',
                icon: 'search',
                status: {
                  pinned: false,
                  default: true
                },
                steps: [
                  {
                    name: 'Enter Search Query',
                    type: 'mandatory',
                    is_completed: false,
                    component: 'form_input',
                    status: 'default',
                    data: '',
                    commands: [
                      { type: 'set_mode', data: { mode: 'search' } },
                      { type: 'set_search_query', data: { query: 'action' } }
                    ]
                  },
                  {
                    name: 'Select Scope',
                    type: 'optional',
                    is_completed: false,
                    component: 'input_test',
                    status: 'default',
                    data: '',
                    commands: [
                      { type: 'set_flag', data: { flag_type: 'selection', value: 'default' } },
                      { type: 'get_selected', data: {} }
                    ]
                  }
                ]
              },
              {
                name: 'Split Tile',
                icon: 'split',
                status: {
                  pinned: false,
                  default: true
                },
                steps: [
                  {
                    name: 'Choose Split Direction',
                    type: 'mandatory',
                    is_completed: false,
                    component: 'form_tile_split_choice',
                    status: 'default',
                    data: ''
                  }
                ]
              }
            ])
          }
        },
        'docs/': {
          'README.md': {
            $ref: 'README.md'
          }
        }
      }
    }
  }
}
