// DOCS/index.js
// --- Main Export ---
// Usage: const docs = DOCS(__filename)(opts.sid)
//        docs.wrap_isolated(handler_string, docContent)
// Admin: Only first caller (root module) gets admin API

module.exports = function DOCS (filename) {
  return function (sid) { return create_context(filename, sid) }
}

const scope = typeof window !== 'undefined' ? window : global

if (!scope.__DOCS_GLOBAL_STATE__) {
  scope.__DOCS_GLOBAL_STATE__ = {
    docs_mode_active: false,
    docs_mode_listeners: [],
    doc_display_callback: null,
    action_registry: new Map(),
    handler_doc_registry: new Map(),
    ephemeral_state: new Map(),
    missing_resource_warnings: new Set()
  }
}

const state = scope.__DOCS_GLOBAL_STATE__
state.handler_doc_registry = state.handler_doc_registry || new Map()
state.ephemeral_state = state.ephemeral_state || new Map()
state.missing_resource_warnings = state.missing_resource_warnings || new Set()

// --- Static Methods (called as DOCS.method()) ---
// Exported via DOCS admin API (only available to first caller)
function set_docs_mode (active) {
  state.docs_mode_active = active
  if (!active) clear_ephemeral_state()
  state.docs_mode_listeners.forEach(listener => listener(active))
}

function get_docs_mode () { return state.docs_mode_active }

function on_docs_mode_change (listener) {
  state.docs_mode_listeners.push(listener)
  return unsubscribe_docs_mode_change

  function unsubscribe_docs_mode_change () { state.docs_mode_listeners = state.docs_mode_listeners.filter(l => l !== listener) }
}

function set_doc_display_handler (callback) { state.doc_display_callback = callback }

function get_actions (sid) {
  const actions = state.action_registry.get(sid) || []
  if (actions.length === 0) throw new Error('DOCS: No actions registered for SID ' + sid)
  return actions
}

function list_registered () { return Array.from(state.action_registry.keys()) }

function get_toc (sid) {
  return {
    actions: state.action_registry.get(sid) || [],
    handlers: state.handler_doc_registry.get(sid) || []
  }
}

// --- Internal Helpers ---

function resolve_handler_doc (handler, doc) {
  if (doc !== undefined) return doc
  if (handler && handler.docs !== undefined) return handler.docs
  if (handler && handler.info !== undefined) return handler.info
  return doc
}

function register_handler_doc (meta) {
  if (meta.doc === undefined || meta.doc === null) return
  const list = state.handler_doc_registry.get(meta.sid) || []
  if (list.some(entry => entry.event_type === meta.event_type && entry.doc === meta.doc)) return
  list.push({ doc: meta.doc, event_type: meta.event_type, component: meta.component })
  state.handler_doc_registry.set(meta.sid, list)
}

function clear_handler_docs (sid) { state.handler_doc_registry.delete(sid) }

function get_ephemeral_state (meta) {
  let store = state.ephemeral_state.get(meta)
  if (!store) { store = {}; state.ephemeral_state.set(meta, store) }
  return store
}

function clear_ephemeral_state () {
  state.ephemeral_state.forEach(store => {
    for (const key of Object.keys(store)) delete store[key]
  })
}

function verify_actions (actions) {
  if (!Array.isArray(actions)) throw new Error('DOCS: Actions must be array')
  actions.forEach(validate_action)

  function validate_action (action, i) {
    if (!action.name || typeof action.name !== 'string') throw new Error(`DOCS: Action[${i}] Invalid 'name'`)
    if (!action.info || typeof action.info !== 'string') throw new Error(`DOCS: Action[${i}] Invalid 'info'`)
    if (!action.icon || typeof action.icon !== 'string') throw new Error(`DOCS: Action[${i}] Invalid 'icon'`)
    if (!action.status || typeof action.status !== 'object') throw new Error(`DOCS: Action[${i}] Invalid 'status'`)
    if (!action.steps || !Array.isArray(action.steps)) throw new Error(`DOCS: Action[${i}] Invalid 'steps'`)
  }
}

async function display_doc (content, sid) {
  let resolved_content = content
  if (typeof content === 'function') {
    resolved_content = await content()
  } else if (content && typeof content.then === 'function') {
    resolved_content = await content
  }

  if (state.doc_display_callback) {
    state.doc_display_callback({ content: resolved_content || 'No documentation available', sid })
  }
}

function create_sys_api (meta) {
  const messages = create_message_api(meta)
  const drive = create_drive_api(meta)
  const sdb = create_sdb_api(meta, drive)

  return {
    is_docs_mode: () => state.docs_mode_active,
    get_doc: () => meta.doc || 'No documentation available',
    get_meta: () => ({ ...meta }),
    show_doc: () => display_doc(meta.doc || 'No documentation available', meta.sid),
    show_action_info: action => show_action_info(meta.sid, resolve_action(meta.sid, action)),
    get_actions: () => state.action_registry.get(meta.sid) || [],
    get_action: name => get_action(meta.sid, name),
    trigger_action: (action, options) => trigger_action(meta, action, options),
    action: (action, options) => trigger_action(meta, action, options),
    send: (channel, type, refs, data) => send_message(meta, channel, type, refs, data),
    _: messages,
    drive,
    sdb,
    state: get_ephemeral_state(meta),
    get_drive: () => drive,
    get_sdb: () => sdb
  }
}

function get_action (sid, name) {
  const actions = state.action_registry.get(sid) || []
  return actions.find(action => action.name === name) || null
}

function resolve_action (sid, action) {
  if (typeof action === 'string') return get_action(sid, action)
  return action || null
}

function trigger_action (meta, action, options = {}) {
  const resolved_action = resolve_action(meta.sid, action)
  if (state.docs_mode_active) return show_action_info(meta.sid, resolved_action)

  if (typeof options.run === 'function') return options.run(resolved_action)
  if (!options.channel || !options.type) {
    warn_once('action:' + meta.sid, 'DOCS: sys.trigger_action requires options.channel and options.type in normal mode')
    return false
  }

  const refs = options.refs || {}
  const data = options.data === undefined ? resolved_action : options.data
  return send_message(meta, options.channel, options.type, refs, data)
}

function get_sys_resources (meta) {
  return meta.resources || {}
}

function create_message_api (meta) {
  return new Proxy({}, {
    get: function get_channel (_, channel) {
      return function send_channel_message (type, refs, data) {
        return send_message(meta, channel, type, refs, data)
      }
    }
  })
}

function send_message (meta, channel, type, refs, data) {
  if (state.docs_mode_active) return false

  const sender = get_sys_resources(meta)._
  if (!sender || typeof sender[channel] !== 'function') {
    warn_once('_.' + channel + ':' + meta.sid, 'DOCS: sys._.' + channel + ' is unavailable; message suppressed')
    return false
  }

  return sender[channel](type, refs, data)
}

function create_drive_api (meta) {
  return {
    get: path => get_drive_file(meta, path),
    put: (path, data) => put_drive_file(meta, path, data)
  }
}

function get_drive_file (meta, path) {
  if (state.docs_mode_active) return Promise.resolve({ raw: null, path })

  const drive = get_sys_resources(meta).drive
  if (!drive || typeof drive.get !== 'function') {
    warn_once('drive.get:' + meta.sid, 'DOCS: sys.drive.get is unavailable; returning empty file')
    return Promise.resolve({ raw: null, path })
  }

  return drive.get(path)
}

function put_drive_file (meta, path, data) {
  if (state.docs_mode_active) return Promise.resolve(false)

  const drive = get_sys_resources(meta).drive
  if (!drive || typeof drive.put !== 'function') {
    warn_once('drive.put:' + meta.sid, 'DOCS: sys.drive.put is unavailable; write suppressed')
    return Promise.resolve(false)
  }

  return drive.put(path, data)
}

function create_sdb_api (meta, drive) {
  return {
    watch: handler => watch_sdb(meta, handler),
    drive
  }
}

function watch_sdb (meta, handler) {
  if (state.docs_mode_active) return Promise.resolve([])

  const sdb = get_sys_resources(meta).sdb
  if (!sdb || typeof sdb.watch !== 'function') {
    warn_once('sdb.watch:' + meta.sid, 'DOCS: sys.sdb.watch is unavailable; returning empty subscriptions')
    return Promise.resolve([])
  }

  return sdb.watch(handler)
}

function warn_once (key, message) {
  if (state.missing_resource_warnings.has(key)) return
  state.missing_resource_warnings.add(key)
  console.warn(message)
}

// --- Instance Methods (called as docs.method()) ---

function wrap (handler, meta = {}, make_sys = create_sys_api, options = {}) {
  meta = { ...meta, doc: get_handler_doc(handler, meta.doc) }
  const run_in_docs_mode = options.run_in_docs_mode === true
  const sys = make_sys(meta)

  return async function wrapped_handler (event) {
    if (sys.is_docs_mode() && !run_in_docs_mode) {
      if (event && event.preventDefault) {
        event.preventDefault()
        event.stopPropagation()
      }
      sys.show_doc()
      return
    }
    return handler.call(this, event, sys)
  }

  function get_handler_doc (handler, doc) {
    if (doc !== undefined) return doc
    if (handler.docs !== undefined) return handler.docs
    if (handler.info !== undefined) return handler.info
    return doc
  }
}

function wrap_isolated (handler_string, meta = {}, options = {}) {
  try {
    const params = 'meta, make_sys'
    const run_in_docs_mode = options.run_in_docs_mode !== false
    const source = `return (${wrap.toString()})(${handler_string}, ${params}, { run_in_docs_mode: ${run_in_docs_mode} })`
    // eslint-disable-next-line no-new-func
    const isolated_fn = new Function(params, source)(meta, create_sys_api)
    return isolated_fn
  } catch (err) {
    console.error('handler function is not allowed to access closure scope', err)
    return wrap(() => {}, meta)
  }
}

function hook (dom, meta = {}) {
  if (!dom) return dom

  const proto = Object.getPrototypeOf(Object.getPrototypeOf(dom))
  if (!proto) return dom

  Object.keys(proto).forEach(hook_event_handler)

  function hook_event_handler (key) {
    if (key.startsWith('on') && typeof dom[key] === 'function') {
      const original = dom[key]
      register_handler_doc({ ...meta, event_type: key, doc: resolve_handler_doc(original, meta.doc) })
      dom[key] = wrap(original, { ...meta, event_type: key })
    }
  }

  return dom
}

// --- Context Factory (creates instance with component scope) ---

function register_actions (sid, actions) {
  verify_actions(actions)
  state.action_registry.set(sid, actions)
}

function show_action_info (sid, action) {
  if (!state.docs_mode_active) return false
  if (!action || typeof action.info !== 'string') return false
  display_doc(action.info, sid)
  return true
}

let admin = true
function create_context (filename, sid) {
  const resources = {}
  const api = {
    wrap: wrap_with_component,
    wrap_isolated: wrap_isolated_with_component,
    hook: hook_with_component,
    get_docs_mode,
    on_docs_mode_change,
    get_toc: () => get_toc(sid),
    clear_handler_docs: () => clear_handler_docs(sid),
    register_actions: register_component_actions,
    show_action_info: show_component_action_info,
    set_sys
  }
  const admin_api = {
    set_docs_mode,
    set_doc_display_handler,
    get_actions,
    get_toc,
    clear_handler_docs,
    list_registered
  }
  const context = admin ? (admin = false, Object.assign({ admin: admin_api }, api)) : api
  return context

  function wrap_with_component (handler, doc) {
    warn_once('deprecated:wrap:' + filename, 'DOCS: docs.wrap is deprecated; use docs.wrap_isolated instead. (' + filename + ')')
    const meta = { doc: resolve_handler_doc(handler, doc), sid, component: filename, resources }
    register_handler_doc(meta)
    return wrap(handler, meta)
  }
  function wrap_isolated_with_component (handler_string, doc, options) {
    const meta = { doc, sid, component: filename, resources }
    register_handler_doc(meta)
    return wrap_isolated(handler_string, meta, options)
  }
  function hook_with_component (dom, doc) { return hook(dom, { doc, sid, component: filename, resources }) }
  function register_component_actions (actions) { return register_actions(sid, actions) }
  function show_component_action_info (action) { return show_action_info(sid, action) }
  function set_sys (api) { Object.assign(resources, api) }
}
