// tabs/tabs.js
const STATE = require('STATE')
const statedb = STATE(__filename)
const { get } = statedb(fallback_module)
const DOCS = require('DOCS')
const net = require('net_helper')

module.exports = component

async function component (opts, invite) {
  const { id, sdb } = await get(opts.sid)
  const { drive } = sdb

  const on = {
    variables: onvariables,
    style: inject,
    icons: iconject,
    scroll: onscroll
  }
  const div = document.createElement('div')
  const shadow = div.attachShadow({ mode: 'closed' })
  shadow.innerHTML = `
  <div class="tab-entries main"></div>`
  const entries = shadow.querySelector('.tab-entries')
  const sheet = new CSSStyleSheet()
  shadow.adoptedStyleSheets = [sheet]

  let init = false
  let variables = []
  let dricons = []
  let active = null
  let tab_group_el = null
  let tab_group_expanded = false
  let tile_is_focused = false
  const default_tabs = {}
  const link_tabs = {}
  const variable_tabs = {}
  let ARROW_LEFT_SVG = ''
  let ARROW_RIGHT_SVG = ''
  const docs = DOCS(__filename)(opts.sid)
  const { io, _ } = net(id)
  docs.set_sys({ _, sdb, drive })

  // Register actions with DOCS system
  const actions_file = await drive.get('actions/commands.json')
  if (actions_file.raw) {
    const actions_data = typeof actions_file.raw === 'string' ? JSON.parse(actions_file.raw) : actions_file.raw
    docs.register_actions(actions_data)
  }

  await sdb.watch(onbatch)
  io.on = {
    up: io_up()
  }
  if (invite) io.accept(invite)
  if (entries) {
    let is_down = false
    let start_x
    let scroll_start

    function stop_drag_scroll () {
      is_down = false
      entries.classList.remove('grabbing')
      update_scroll_position()
    }

    function move_drag_scroll (pointer_x) {
      if (!is_down) return
      if (entries.scrollWidth <= entries.clientWidth) return stop_drag_scroll()
      entries.scrollLeft = scroll_start - (pointer_x - start_x) * 1.5
    }

    entries.onmousedown = on_entries_mousedown

    function on_entries_mousedown (e) {
      if (entries.scrollWidth <= entries.clientWidth) return
      is_down = true
      entries.classList.add('grabbing')
      start_x = e.pageX - entries.offsetLeft
      scroll_start = entries.scrollLeft
      window.onmousemove = on_window_mousemove
      window.onmouseup = on_window_mouseup
    }

    function on_window_mousemove (e) {
      move_drag_scroll(e.pageX - entries.offsetLeft)
      e.preventDefault()
    }

    function on_window_mouseup () {
      stop_drag_scroll()
      window.onmousemove = null
      window.onmouseup = null
    }

    entries.onmouseleave = stop_drag_scroll
    entries.ontouchstart = on_entries_touchstart

    function on_entries_touchstart (e) {
      if (entries.scrollWidth <= entries.clientWidth) return
      is_down = true
      start_x = e.touches[0].pageX - entries.offsetLeft
      scroll_start = entries.scrollLeft
    }

    ;['ontouchend', 'ontouchcancel'].forEach(bind_touch_end_handler)
    entries.ontouchmove = on_entries_touchmove

    function bind_touch_end_handler (event_name) { entries[event_name] = stop_drag_scroll }
    function on_entries_touchmove (e) {
      move_drag_scroll(e.touches[0].pageX - entries.offsetLeft)
      e.preventDefault()
    }
  }
  return div

  function io_up () {
    const on = {
      add_link_tab: handle_add_link_tab,
      remove_link_tab: handle_remove_link_tab,
      add_default_tab: handle_add_default_tab,
      restore_tab: handle_restore_tab,
      show_collapsed_tab_group: handle_show_collapsed_tab_group,
      hide_collapsed_tab_group: handle_hide_collapsed_tab_group,
      tile_focus_changed: handle_tile_focus_changed
    }
    return function onmessage (msg) {
      console.error('tabs: message from up', msg)
      ;(on[msg.type] || onfail)(msg)
    }
    function handle_add_link_tab ({ data }) { add_link_tab(data) }
    function handle_remove_link_tab ({ data }) { remove_link_tab(data) }
    function handle_add_default_tab ({ data }) { add_default_tab(data) }
    function handle_restore_tab ({ data }) { create_btn({ name: data.name, id: data.id }) }
    function handle_show_collapsed_tab_group ({ data }) { show_collapsed_tab_group(data) }
    function handle_hide_collapsed_tab_group () { hide_collapsed_tab_group() }
    function handle_tile_focus_changed ({ data }) { update_tab_focus_state(data) }
    function onfail (msg) { console.error('tabs: unknown message', msg) }
  }

  function add_default_tab ({ name, program, tile_id, id }) {
    const tab_id = id || `tab_${Date.now()}`
    if (default_tabs[tab_id]) {
      console.error('tabs: default tab already exists', tab_id)
      return
    }
    const el = document.createElement('div')
    el.innerHTML = `
    <span class="icon">${dricons[1] || '📄'}</span>
    <span class='name'>${tab_id}</span>
    <span class="name">${name || 'New Tab'}</span>
    <button class="btn">${dricons[0] || '×'}</button>`
    el.className = 'tabsbtn default-tab active'
    const name_el = el.querySelector('.name')
    const close_btn = el.querySelector('.btn')
    default_tabs[tab_id] = { el, name_el, close_btn, name, program }
    if (active) active.classList.remove('active')
    active = el
    name_el.onclick = switch_active
    close_btn.onclick = close_tab
    entries.appendChild(el)
    console.error('tabs: default tab added', tab_id)

    function switch_active () {
      console.error('tabs: default tab clicked', tab_id)
      if (active) active.classList.remove('active')
      el.classList.add('active')
      active = el
      const data = { id: tab_id, name, program }
      _.up('ui_focus', {}, { type: 'tab', sid: opts.sid })
      _.up('tab_name_clicked', {}, data)
    }
    function close_tab (e) {
      e.stopPropagation()
      console.error('tabs: default tab close clicked', tab_id)
      el.remove()
      delete default_tabs[tab_id]
      if (active === el) active = null
      _.up('tab_close_clicked', {}, { id: tab_id, name })
      if (Object.keys(default_tabs).length === 0) {
        _.up('all_tabs_closed', {}, null)
      }
    }
  }

  function add_link_tab ({ tile_id, name, direction }) {
    const link_tab_id = `split_tile_${tile_id}`
    if (link_tabs[link_tab_id]) {
      console.error('tabs: link tab already exists', link_tab_id)
      return
    }
    const el = document.createElement('div')
    el.innerHTML = `
    <span class="icon">⊞</span>
    <span class="name">${name || 'Split ' + direction}</span>`
    el.className = 'tabsbtn link-tab'
    const name_el = el.querySelector('.name')
    link_tabs[link_tab_id] = { el, name_el, tile_id, name, direction }
    el.onclick = on_link_tab_click
    entries.appendChild(el)
    console.error('tabs: link tab added', link_tab_id)

    function on_link_tab_click () {
      console.error('tabs: link tab clicked', link_tab_id)
      _.up('link_tab_clicked', {}, { tile_id, link_tab_id })
    }
  }

  function remove_link_tab ({ tile_id }) {
    const link_tab_id = `split_tile_${tile_id}`
    const tab = link_tabs[link_tab_id]
    if (tab) {
      tab.el.remove()
      delete link_tabs[link_tab_id]
      console.error('tabs: link tab removed', link_tab_id)
    }
  }

  // ---------------------------
  // COLLAPSED TAB GROUP
  // Shows collapsed tile's tabs inline (like Chrome tab groups)
  // ---------------------------

  function show_collapsed_tab_group ({ tiles }) {
    if (tab_group_el) hide_collapsed_tab_group()

    // Container holds the group tab + inline child tabs
    tab_group_el = document.createElement('div')
    tab_group_el.className = 'tab-group-inline'

    // Group tab — styled same as default tabs
    const group_tab = document.createElement('div')
    group_tab.className = 'tabsbtn tab-group-tab'
    const direction = (tiles[0] && tiles[0].direction) || 'right'
    const icon_svg = (direction === 'left' || direction === 'up') ? ARROW_LEFT_SVG : ARROW_RIGHT_SVG
    group_tab.innerHTML = `<span class="icon">${icon_svg}</span><span class="name">Split</span>`
    group_tab.onclick = toggle_tab_group_expand
    tab_group_el.appendChild(group_tab)

    // Child tabs — the actual tabs from the collapsed tile, shown inline
    const children_container = document.createElement('div')
    children_container.className = 'tab-group-children'

    for (const tile of tiles) {
      const tile_tabs = tile.tabs || []
      for (const tab of tile_tabs) {
        const tab_el = document.createElement('div')
        tab_el.className = 'tabsbtn tab-group-child'
        tab_el.innerHTML = `<span class="icon">${dricons[1] || '📄'}</span><span class="name">${tab.name || 'Tab'}</span>`
        tab_el.onclick = on_child_tab_click(tile.tile_id)
        children_container.appendChild(tab_el)
      }
      // If no tabs tracked, show at least one entry
      if (tile_tabs.length === 0) {
        const tab_el = document.createElement('div')
        tab_el.className = 'tabsbtn tab-group-child'
        tab_el.innerHTML = `<span class="icon">${dricons[1] || '📄'}</span><span class="name">New Tab</span>`
        tab_el.onclick = on_child_tab_click(tile.tile_id)
        children_container.appendChild(tab_el)
      }
    }

    tab_group_el.appendChild(children_container)
    entries.insertBefore(tab_group_el, entries.firstChild)
    console.error('tabs: collapsed tab group shown')

    function toggle_tab_group_expand () {
      tab_group_expanded = !tab_group_expanded
      tab_group_el.classList.toggle('expanded', tab_group_expanded)
    }

    function on_child_tab_click (tile_id) {
      return function () {
        console.error('tabs: tab group child clicked, requesting expand', tile_id)
        _.up('tab_group_tile_clicked', {}, { tile_id })
      }
    }
  }

  function hide_collapsed_tab_group () {
    if (tab_group_el) {
      tab_group_el.remove()
      tab_group_el = null
      tab_group_expanded = false
      console.error('tabs: collapsed tab group hidden')
    }
  }

  // ---------------------------
  // FOCUS STATE
  // Updates active tab styling based on tile focus:
  // - Tile in focus: active tab gets highlight background
  // - Tile NOT in focus: active tab gets highlight border only
  // ---------------------------

  function update_tab_focus_state ({ is_focused }) {
    entries.classList.toggle('tile-focused', is_focused)
    entries.classList.toggle('tile-inactive', !is_focused)
  }

  async function create_btn ({ name, id }, index = 0) {
    if (variable_tabs[id]) {
      console.error('tabs: variable tab already exists', id)
      return
    }
    const el = document.createElement('div')
    el.innerHTML = `
    <span class="icon">${dricons[index + 1] || dricons[1] || '📄'}</span>
    <span class='name'>${id}</span>
    <span class="name">${name}</span>
    <button class="btn">${dricons[0] || '×'}</button>`

    el.className = 'tabsbtn'
    const name_el = el.querySelector('.name')
    const close_btn = el.querySelector('.btn')

    name_el.draggable = false

    name_el.__tab = { id, name }
    name_el.onclick = docs.wrap_isolated(
      'function (event, sys) { if (sys.is_docs_mode()) return sys.show_doc(); var t = event.currentTarget.__tab; sys._.up("ui_focus", {}, { type: "tab", sid: sys.get_meta().sid }); sys._.up("tab_name_clicked", {}, t) }',
      get_doc_content
    )
    close_btn.__tab = { id, name }
    close_btn.__variable_tabs = variable_tabs
    close_btn.onclick = docs.wrap_isolated(
      'function (event, sys) { if (sys.is_docs_mode()) return sys.show_doc(); var el = event.currentTarget.parentElement, t = event.currentTarget.__tab; event.stopPropagation(); el.remove(); delete event.currentTarget.__variable_tabs[t.id]; sys._.up("ui_focus", {}, { type: "tab", sid: sys.get_meta().sid }); sys._.up("tab_close_clicked", {}, t) }',
      get_doc_content
    )
    variable_tabs[id] = { el, name }

    async function get_doc_content () {
      const doc_file = await drive.get('docs/README.md')
      return doc_file.raw || 'No documentation available'
    }

    entries.appendChild(el)
  }

  async function onbatch (batch) {
    for (const { type, paths } of batch) {
      const data = await Promise.all(paths.map(load_path_raw))
      const func = on[type] || fail
      func(data, type)
    }
    if (!init) {
      if (!opts.ids) variables.forEach(create_btn)
      init = true
    } else {
      // TODO: Here we can handle drive updates
    }

    function load_path_raw (path) { return drive.get(path).then(read_drive_file_raw) }
    function read_drive_file_raw (file) { return file.raw }
  }
  function fail (data, type) { console.warn('invalid message', { cause: { data, type } }) }
  function inject (data) { sheet.replaceSync(data[0]) }

  function onvariables (data) {
    const vars = typeof data[0] === 'string' ? JSON.parse(data[0]) : data[0]
    variables = vars
  }

  function iconject (data) {
    dricons = data
    ARROW_LEFT_SVG = data[4] || ''
    ARROW_RIGHT_SVG = data[5] || ''
  }

  function update_scroll_position () {
    // TODO
  }

  function onscroll (data) {
    setTimeout(apply_scroll_position, 200)
    function apply_scroll_position () {
      if (entries) {
        entries.scrollLeft = data
      }
    }
  }
}

function fallback_module () {
  return {
    api: fallback_instance,
    _: {
      DOCS: {
        $: ''
      },
      net_helper: {
        $: ''
      }
    }
  }
  function fallback_instance () {
    return {
      _: {
        DOCS: {
          0: ''
        },
        net_helper: {
          0: ''
        }
      },
      drive: {
        'icons/': {
          'cross.svg': {
            $ref: 'cross.svg'
          },
          '1.svg': {
            $ref: 'icon.svg'
          },
          '2.svg': {
            $ref: 'icon.svg'
          },
          '3.svg': {
            $ref: 'icon.svg'
          },
          '4.svg': {
            $ref: 'arrow-left.svg'
          },
          '5.svg': {
            $ref: 'arrow-right.svg'
          }
        },
        'actions/': {
          'commands.json': {
            raw: JSON.stringify([
              {
                name: 'New Tab',
                info: 'Create a new tab in the current tab strip.',
                icon: 'plus',
                status: {
                  pinned: true,
                  default: true
                },
                steps: [
                  { name: 'Enter Tab Name', type: 'optional', is_completed: false, component: 'form_input', status: 'default', data: '' }
                ]
              },
              {
                name: 'Duplicate Tab',
                info: 'Copy an existing tab and open the duplicate as a new tab.',
                icon: 'copy',
                status: {
                  pinned: false,
                  default: false
                },
                steps: [
                  { name: 'Select Tab to Duplicate', type: 'mandatory', is_completed: false, component: 'form_input', status: 'default', data: '' },
                  { name: 'Enter New Tab Name', type: 'optional', is_completed: false, component: 'form_input', status: 'default', data: '' }
                ]
              },
              {
                name: 'Close Tab',
                info: 'Close the selected tab after confirmation.',
                icon: 'close',
                status: {
                  pinned: false,
                  default: true
                },
                steps: [
                  { name: 'Select Tab to Close', type: 'mandatory', is_completed: false, component: 'form_input', status: 'default', data: '' },
                  { name: 'Confirm Close', type: 'mandatory', is_completed: false, component: 'form_input', status: 'default', data: '' }
                ]
              }
            ])
          }
        },
        'variables/': {
          'tabs.json': {
            $ref: 'tabs.json'
          }
        },
        'scroll/': {
          'position.json': {
            raw: '100'
          }
        },
        'docs/': {
          'README.md': {
            $ref: 'README.md'
          }
        },
        'style/': {
          'theme.css': {
            $ref: 'style.css'
          }
        }
      }
    }
  }
}
