// console_history/console_history.js
const STATE = require('STATE')
const statedb = STATE(__filename)
const { get } = statedb(fallback_module)
const DOCS = require('DOCS')
const DEBUG = require('DEBUG')
const net = require('net_helper')

module.exports = console_history

async function console_history (opts, invite) {
  const { id, sdb } = await get(opts.sid)
  const { drive } = sdb

  const on = {
    style: inject,
    commands: oncommands,
    icons: iconject
  }

  const el = document.createElement('div')
  const shadow = el.attachShadow({ mode: 'closed' })
  shadow.innerHTML = `
  <div class="console-history-container main">
    <div class="console-menu">
      <div class="commands-list"></div>
    </div>
    <div class="console-search">
      <span class="console-search-icon"><svg width="14" height="14" viewBox="0 0 16 16" fill="none" xmlns="http://www.w3.org/2000/svg"><circle cx="7" cy="7" r="5" stroke="currentColor" stroke-width="1.5"/><path d="M11 11L14 14" stroke="currentColor" stroke-width="1.5" stroke-linecap="round"/></svg></span>
      <input class="console-search-input" type="text" placeholder="search input" />
      <button class="console-search-btn console-search-clear" title="Clear search"><svg width="14" height="14" viewBox="0 0 16 16" fill="none" xmlns="http://www.w3.org/2000/svg"><path d="M4 4L12 12M12 4L4 12" stroke="currentColor" stroke-width="1.5" stroke-linecap="round"/></svg></button>
      <button class="console-search-btn console-search-filter" title="Toggle default (demo) entries"><svg width="14" height="14" viewBox="0 0 16 16" fill="none" xmlns="http://www.w3.org/2000/svg"><path d="M2 3H14L9.5 8.5V13L6.5 11.5V8.5L2 3Z" stroke="currentColor" stroke-width="1.5" stroke-linejoin="round"/></svg></button>
    </div>
  </div>`
  const sheet = new CSSStyleSheet()
  shadow.adoptedStyleSheets = [sheet]
  const commands_list = shadow.querySelector('.commands-list')
  const search_input = shadow.querySelector('.console-search-input')
  const clear_btn = shadow.querySelector('.console-search-clear')
  const filter_btn = shadow.querySelector('.console-search-filter')
  search_input.oninput = on_search_input
  clear_btn.onclick = on_clear_search
  filter_btn.onclick = on_toggle_default_entries

  let default_commands = []
  const live_commands = []
  let search_term = ''
  let dricons = []
  const docs = DOCS(__filename)(opts.sid)
  const debug = DEBUG(__filename)(opts.sid)
  const { io, _ } = net(id)
  docs.set_sys({ _, sdb, drive })
  debug.on_change(on_debug_change)

  // Register actions with DOCS system
  const actions_file = await drive.get('actions/commands.json')
  if (actions_file.raw) {
    const actions_data = typeof actions_file.raw === 'string' ? JSON.parse(actions_file.raw) : actions_file.raw
    docs.register_actions(actions_data)
  } else {
    console.error('actions.json not found')
  }

  await sdb.watch(onbatch)
  io.on = {
    up: io_up()
  }
  if (invite) io.accept(invite)
  return el

  function io_up () {
    const on_message = {
      record_closed_tab: handle_record_closed_tab
    }
    return function onmessage (msg) {
      const handler = on_message[msg.type] || onmessage_fail
      handler(msg)
    }
    function handle_record_closed_tab (msg) { record_closed_tab(msg.data) }
    function onmessage_fail (msg) { console.warn('console_history: unknown message', { cause: msg && msg.type }) }
  }

  function record_closed_tab (data) {
    const tab = data || {}
    const entry = {
      icon_type: 'file',
      name_path: tab.name || tab.id || 'Tab',
      label: 'closed',
      pending: true,
      status: null,
      restore_data: { id: tab.id, name: tab.name }
    }
    live_commands.unshift(entry)
    render_commands()
  }

  function resolve_closed_tab (entry, status, do_restore) {
    entry.pending = false
    entry.status = status
    entry.label = status
    if (do_restore) _.up('restore_tab', {}, entry.restore_data || { name: entry.name_path })
    render_commands()
  }

  function on_search_input (e) {
    search_term = e.target.value
    render_commands()
  }

  function on_clear_search () {
    search_term = ''
    search_input.value = ''
    render_commands()
  }

  function on_toggle_default_entries () {
    debug.set_flag('show_default_entries', !debug.get_flag('show_default_entries'))
  }

  function on_debug_change () { render_commands() }

  function create_command_item (command_data) {
    const command_el = document.createElement('div')
    command_el.className = 'command-item'

    const icon_html = dricons[command_data.icon_type] || dricons.file || ''
    const is_pending = !!command_data.pending
    const status = command_data.status

    let right_html = ''
    if (is_pending) {
      right_html =
        '<div class="action-icon restore-action" title="Restore">' + (dricons.restore || '') + '</div>' +
        '<div class="action-icon delete-action" title="Delete">' + (dricons.delete || '') + '</div>'
    } else if (status) {
      const status_class = String(status).toLowerCase().indexOf('delet') === 0 ? 'deleted' : 'restored'
      right_html = '<span class="status-text status-' + status_class + '">' + status + '</span>'
    }

    command_el.innerHTML = `
    <div class="command-content">
      <div class="command-icon">${icon_html}</div>
      <div class="command-info">
        <div class="command-name">${command_data.name_path}</div>
        <div class="command-label">${command_data.label || ''}</div>
      </div>
      ${right_html ? `<div class="command-actions">${right_html}</div>` : ''}
    </div>`

    command_el.__command = command_data
    command_el.onclick = docs.wrap_isolated(
      'function (event, sys) { if (sys.is_docs_mode()) return sys.show_doc(); var el = event.currentTarget; var previous = el.parentElement.querySelector(".command-item.selected"); if (previous) previous.classList.remove("selected"); el.classList.add("selected"); sys._.up("ui_focus", {}, { type: "command_history", sid: sys.get_meta().sid }); sys._.up("command_clicked", {}, el.__command) }',
      get_doc_content
    )

    const restore_el = command_el.querySelector('.restore-action')
    const delete_el = command_el.querySelector('.delete-action')
    if (restore_el) restore_el.onclick = on_restore_click
    if (delete_el) delete_el.onclick = on_delete_click

    function on_restore_click (e) {
      e.stopPropagation()
      resolve_closed_tab(command_data, 'Restored', true)
    }

    function on_delete_click (e) {
      e.stopPropagation()
      resolve_closed_tab(command_data, 'Deleted', false)
    }

    async function get_doc_content () {
      const doc_file = await drive.get('docs/README.md')
      return doc_file.raw || 'No documentation available'
    }

    return command_el
  }
  function render_commands () {
    commands_list.replaceChildren()
    const show_defaults = debug.get_flag('show_default_entries')
    const base = show_defaults ? live_commands.concat(default_commands) : live_commands.slice()
    const term = search_term.trim().toLowerCase()
    const visible = term ? base.filter(matches_search) : base
    visible.forEach(append_command_item)

    function matches_search (command) {
      const haystack = [command.name_path, command.label, command.status]
        .filter(Boolean)
        .join(' ')
        .toLowerCase()
      return haystack.includes(term)
    }

    function append_command_item (command, index) {
      const command_item = create_command_item(command, index)
      commands_list.appendChild(command_item)
    }
  }
  async function onbatch (batch) {
    for (const { type, paths } of batch) {
      const data = await Promise.all(paths.map(load_path_raw))
      const func = on[type] || fail
      func(data, type)
    }
    render_commands()

    function load_path_raw (path) { return drive.get(path).then(read_drive_file_raw) }
    function read_drive_file_raw (file) { return file.raw }
  }

  function fail (data, type) { console.warn('invalid message', { cause: { data, type } }) }

  function inject (data) { sheet.replaceSync(data[0]) }
  function oncommands (data) {
    const commands_data = typeof data[0] === 'string' ? JSON.parse(data[0]) : data[0]
    default_commands = Array.isArray(commands_data) ? commands_data.map(normalize_default) : []
  }

  function normalize_default (entry) {
    return {
      icon_type: entry.icon_type || 'file',
      name_path: entry.name_path || entry.name || 'entry',
      label: entry.label || entry.status || entry.command || '',
      pending: !!entry.pending,
      status: entry.pending ? null : (entry.status || null),
      restore_data: entry.restore_data || { name: entry.name_path }
    }
  }

  function iconject (data) {
    dricons = {
      file: data[0] || '',
      bulb: data[1] || '',
      restore: data[2] || '',
      delete: data[3] || ''
    }
  }
}

function fallback_module () {
  return {
    api: fallback_instance,
    _: {
      DOCS: {
        $: ''
      },
      DEBUG: {
        $: ''
      },
      net_helper: {
        $: ''
      }
    }
  }

  function fallback_instance () {
    return {
      _: {
        DOCS: {
          0: ''
        },
        DEBUG: {
          0: ''
        },
        net_helper: {
          0: ''
        }
      },
      drive: {
        'commands/': {
          'list.json': {
            $ref: 'commands.json'
          }
        },
        'actions/': {
          'commands.json': {
            raw: JSON.stringify([
              {
                name: 'Clear History',
                info: 'Clear the stored console history after confirmation.',
                icon: 'trash',
                status: {
                  pinned: false,
                  default: true
                },
                steps: [
                  { name: 'Confirm Clear', type: 'mandatory', is_completed: false, component: 'form_input', status: 'default', data: '' }
                ]
              },
              {
                name: 'Export History',
                info: 'Export console history to the selected format and location.',
                icon: 'download',
                status: {
                  pinned: true,
                  default: false
                },
                steps: [
                  { name: 'Choose Format', type: 'mandatory', is_completed: false, component: 'form_input', status: 'default', data: '' },
                  { name: 'Select Location', type: 'mandatory', is_completed: false, component: 'form_input', status: 'default', data: '' }
                ]
              },
              {
                name: 'Search History',
                info: 'Search through recorded console history entries.',
                icon: 'search',
                status: {
                  pinned: false,
                  default: true
                },
                steps: [
                  { name: 'Enter Search Term', type: 'mandatory', is_completed: false, component: 'form_input', status: 'default', data: '' }
                ]
              }
            ])
          }
        },
        'icons/': {
          'file.svg': {
            raw: `<svg width="16" height="16" viewBox="0 0 16 16" fill="none" xmlns="http://www.w3.org/2000/svg">
              <path d="M9.5 1H3.5C3.10218 1 2.72064 1.15804 2.43934 1.43934C2.15804 1.72064 2 2.10218 2 2.5V13.5C2 13.8978 2.15804 14.2794 2.43934 14.5607C2.72064 14.8420 3.10218 15 3.5 15H12.5C12.8978 15 13.2794 14.8420 13.5607 14.5607C13.8420 14.2794 14 13.8978 14 13.5V5.5L9.5 1Z" stroke="currentColor" stroke-width="1.5" stroke-linecap="round" stroke-linejoin="round"/>
              <path d="M9.5 1V5.5H14" stroke="currentColor" stroke-width="1.5" stroke-linecap="round" stroke-linejoin="round"/>
            </svg>`
          },
          'bulb.svg': {
            raw: `<svg width="16" height="16" viewBox="0 0 16 16" fill="none" xmlns="http://www.w3.org/2000/svg">
              <path d="M8 1C6.4087 1 4.88258 1.63214 3.75736 2.75736C2.63214 3.88258 2 5.4087 2 7C2 8.5913 2.63214 10.1174 3.75736 11.2426C4.88258 12.3679 6.4087 13 8 13C9.5913 13 11.1174 12.3679 12.2426 11.2426C13.3679 10.1174 14 8.5913 14 7C14 5.4087 13.3679 3.88258 12.2426 2.75736C11.1174 1.63214 9.5913 1 8 1Z" stroke="currentColor" stroke-width="1.5" stroke-linecap="round" stroke-linejoin="round"/>
              <path d="M6.5 14H9.5" stroke="currentColor" stroke-width="1.5" stroke-linecap="round" stroke-linejoin="round"/>
            </svg>`
          },
          'restore.svg': {
            raw: `<svg xmlns="http://www.w3.org/2000/svg" width="16" height="16" fill="currentColor" class="bi bi-arrow-counterclockwise" viewBox="0 0 16 16">
              <path fill-rule="evenodd" d="M8 3a5 5 0 1 1-4.546 2.914.5.5 0 0 0-.908-.417A6 6 0 1 0 8 2v1z"/>
              <path d="M8 4.466V.534a.25.25 0 0 0-.41-.192L5.23 2.308a.25.25 0 0 0 0 .384l2.36 1.966A.25.25 0 0 0 8 4.466z"/>
            </svg>`
          },
          'delete.svg': {
            raw: `<svg xmlns="http://www.w3.org/2000/svg" width="16" height="16" fill="currentColor" class="bi bi-trash3" viewBox="0 0 16 16">
              <path d="M6.5 1h3a.5.5 0 0 1 .5.5v1H6v-1a.5.5 0 0 1 .5-.5ZM11 2.5v-1A1.5 1.5 0 0 0 9.5 0h-3A1.5 1.5 0 0 0 5 1.5v1H2.506a.58.58 0 0 0-.01 0H1.5a.5.5 0 0 0 0 1h.538l.853 10.66A2 2 0 0 0 4.885 16h6.23a2 2 0 0 0 1.994-1.84l.853-10.66h.538a.5.5 0 0 0 0-1h-.995a.59.59 0 0 0-.01 0H11Zm1.958 1-.846 10.58a1 1 0 0 1-.997.92H4.885a1 1 0 0 1-.997-.92L3.042 3.5h9.916Zm-7.487 1a.5.5 0 0 1 .528.47l.5 8.5a.5.5 0 0 1-.998.06L5 5.03a.5.5 0 0 1 .47-.528ZM8 4.5a.5.5 0 0 1 .5.5v8.5a.5.5 0 0 1-1 0V5a.5.5 0 0 1 .5-.5Zm2.522.47a.5.5 0 0 1 .528.47l-.5 8.5a.5.5 0 1 1-.998-.06l.5-8.5a.5.5 0 0 1 .47-.528Z"/>
            </svg>`
          }
        },
        'docs/': {
          'README.md': {
            $ref: 'README.md'
          }
        },
        'style/': {
          'theme.css': {
            raw: `
              .console-history-container {
                display: flex;
                flex-direction: column;
                flex: 1 1 auto;
                width: 100%;
                height: 100%;
                background: #202124;
                border: 1px solid #3c3c3c;
                box-sizing: border-box;
                box-shadow: 0 4px 12px rgba(0, 0, 0, 0.4);
                z-index: 1;
                overflow: hidden;
                color: #e8eaed;
              }

              .console-menu {
                flex: 1 1 auto;
                min-height: 0;
                overflow-y: auto;
                padding: 0px;
              }

              /* Sticky search bar pinned to the bottom; scrolling the list above
                 it does not move it. */
              .console-search {
                flex: 0 0 auto;
                display: flex;
                align-items: center;
                gap: 8px;
                padding: 8px 12px;
                background: #18191b;
                border-top: 1px solid #3c3c3c;
              }

              .console-search-icon {
                display: flex;
                align-items: center;
                justify-content: center;
                color: #969ba1;
                flex: 0 0 auto;
              }

              .console-search-input {
                flex: 1 1 auto;
                min-width: 0;
                box-sizing: border-box;
                padding: 6px 8px;
                background: transparent;
                color: #e8eaed;
                border: none;
                outline: none;
                font-size: 13px;
              }

              .console-search-input::placeholder { color: #6b7077; }

              .console-search-btn {
                display: flex;
                align-items: center;
                justify-content: center;
                width: 26px;
                height: 26px;
                padding: 0;
                background: transparent;
                border: none;
                color: #969ba1;
                cursor: pointer;
                border-radius: 4px;
              }

              .console-search-btn:hover {
                color: #e8eaed;
                background: rgba(255, 255, 255, 0.08);
              }

              .commands-list {
                display: flex;
                flex-direction: column;
                gap: 0px;
              }

              .command-item {
                display: flex;
                align-items: center;
                padding: 8px 14px;
                background: transparent;
                border-bottom: 1px solid #3c3c3c;
                cursor: pointer;
                transition: background-color 0.15s ease;
              }

              .command-item:last-child {
                border-bottom: none;
              }

              .command-item:hover {
                background: #282a2d;
              }

              .command-item.selected {
                background: #f56300;
              }

              .command-item.selected .command-name,
              .command-item.selected .command-label,
              .command-item.selected .status-text,
              .command-item.selected .action-icon {
                color: #fff;
              }

              .command-content {
                display: flex;
                align-items: center;
                width: 100%;
                gap: 12px;
              }

              .command-icon {
                display: flex;
                align-items: center;
                justify-content: center;
                width: 28px;
                height: 28px;
                border-radius: 6px;
                background: #2b2c2f;
                color: #c8ccd2;
                flex: 0 0 auto;
              }

              .command-item.selected .command-icon {
                background: rgba(255, 255, 255, 0.18);
                color: #fff;
              }

              .command-icon svg {
                width: 16px;
                height: 16px;
              }

              .command-info {
                display: flex;
                flex-direction: column;
                gap: 2px;
                min-width: 0;
                flex: 1 1 auto;
              }

              .command-label {
                font-size: 11px;
                color: #969ba1;
                white-space: nowrap;
                overflow: hidden;
                text-overflow: ellipsis;
              }

              .status-text {
                font-size: 12px;
                white-space: nowrap;
              }

              .status-text.status-restored { color: #46c95d; }
              .status-text.status-deleted { color: #ff6b6b; }

              .console-menu::-webkit-scrollbar {
                width: 10px;
              }
              .console-menu::-webkit-scrollbar-track {
                background: transparent;
              }
              .console-menu::-webkit-scrollbar-thumb {
                background: #30363d;
                border-radius: 999px;
                background-clip: content-box;
                border: 2px solid transparent;
              }
              .console-menu::-webkit-scrollbar-thumb:hover {
                background: #484f58;
              }
              .console-menu {
                scrollbar-width: thin;
                scrollbar-color: #30363d transparent;
              }
              .command-name {
                font-size: 13px;
                font-weight: 400;
                color: #e8eaed;
                white-space: nowrap;
                overflow: hidden;
                text-overflow: ellipsis;
              }

              .command-path {
                font-size: 13px;
                color: #969ba1;
                white-space: nowrap;
                overflow: hidden;
                text-overflow: ellipsis;
              }
              
              .command-separator {
                color: #969ba1;
                margin: 0 4px;
                font-size: 13px;
              }

              .linked-info {
                display: flex;
                align-items: center;
                gap: 6px;
                flex-grow: 1; /* Allow info to take available space */

              }

              .linked-icon {
                display: flex;
                align-items: center;
                justify-content: center;
                width: 16px;
                height: 16px;
                color: #fbbc04; 
              }

              .linked-icon svg {
                width: 14px;
                height: 14px;
              }

              .linked-name {
                font-size: 13px;
                color: #fbbc04;
                font-weight: 400;
                white-space: nowrap;
              }

              .command-actions {
                display: flex;
                align-items: center;
                gap: 10px; /* Adjusted gap */
                margin-left: auto; /* Pushes actions to the right */
              }

              .action-text {
                font-size: 13px;
                color: #969ba1;
                white-space: nowrap;
              }

              .action-icon {
                display: flex;
                align-items: center;
                justify-content: center;
                width: 20px;
                height: 20px;
                color: #969ba1;
                cursor: pointer;
              }

              .action-icon:hover {
                color: #e8eaed;
              }

              .action-icon svg {
                width: 16px;
                height: 16px;
              }
            `
          }
        }
      }
    }
  }
}
