// form_click_rate_test/form_click_rate_test.js
const STATE = require('STATE')
const statedb = STATE(__filename)
const { get } = statedb(fallback_module)
const DOCS = require('DOCS')
const net = require('net_helper')

module.exports = form_click_rate_test

async function form_click_rate_test (opts, invite) {
  const { id, sdb } = await get(opts.sid)
  const { drive } = sdb

  const on = {
    style: inject,
    data: ondata
  }

  const docs = DOCS(__filename)(opts.sid)
  const { io, _ } = net(id)
  docs.set_sys({ _, sdb, drive })
  const milliseconds_action = {
    name: 'Click Rate Result',
    info: 'Calculate and submit the milliseconds taken to complete 10 clicks.',
    icon: 'timer',
    status: {
      pinned: false,
      default: false
    },
    steps: []
  }

  const el = document.createElement('div')
  const shadow = el.attachShadow({ mode: 'closed' })
  shadow.innerHTML = `
  <div class="click-rate-test">
    <div class="title">Click Rate Test</div>
    <button class="click-btn" type="button">
      <span class="count">0</span>
      <span class="label">clicks</span>
    </button>
    <div class="hint">Click as fast as you can. The 10th click completes the action.</div>
    <div class="result"></div>
  </div>
  <style></style>
  `
  const style = shadow.querySelector('style')
  const button = shadow.querySelector('.click-btn')
  const count_el = shadow.querySelector('.count')
  const result_el = shadow.querySelector('.result')

  button.__action = milliseconds_action
  button.__accessible = true
  button.__step_index = 0
  button.__count = 0
  button.__start = 0
  button.onclick = docs.wrap_isolated(
    'function (event, sys) { var btn = event.currentTarget; var count_el = btn.querySelector(".count"); var result_el = btn.parentElement.querySelector(".result"); if (sys.is_docs_mode()) { var s = sys.state; s.count = (s.count || 0) + 1; count_el.textContent = s.count; if (s.count < 10) { sys.show_doc(); return } sys.show_action_info(btn.__action); return } if (!btn.__accessible) return; if (btn.__count === 0) btn.__start = Date.now(); btn.__count += 1; count_el.textContent = btn.__count; if (btn.__count < 10) { sys.drive.put("data/form_click_rate_test.json", { click_count: btn.__count, result: "" }); sys._.up("action_incomplete", {}, { value: btn.__count, index: btn.__step_index }); return } var result = "10 clicks in " + (Date.now() - btn.__start) + " ms"; result_el.textContent = result; sys.drive.put("data/form_click_rate_test.json", { click_count: btn.__count, result: result }); sys._.up("action_submitted", {}, { value: result, index: btn.__step_index }); sys._.up("action_complete", {}, { value: result }) }',
    'Record one click in the 10-click sequence. This event is not an action until the 10th click.'
  )
  docs.register_actions([milliseconds_action])
  docs.on_docs_mode_change(on_docs_mode_change)

  await sdb.watch(onbatch)

  const parent_handler = {
    step_data,
    reset_data
  }

  io.on = {
    up: io_up()
  }
  if (invite) io.accept(invite)

  return el

  function io_up () {
    return function onmessage ({ type, data }) {
      const handler = parent_handler[type] || fail
      handler(data, type)
    }
  }

  async function onbatch (batch) {
    for (const { type, paths } of batch) {
      const data = await Promise.all(paths.map(load_path_raw))
      const func = on[type] || fail
      func(data, type)
    }

    function load_path_raw (path) { return drive.get(path).then(read_drive_file_raw) }
    function read_drive_file_raw (file) { return file.raw }
  }

  function fail (data, type) { console.warn('invalid message', { cause: { data, type } }) }

  function inject (data) {
    style.replaceChildren(create_style_element())

    function create_style_element () {
      const style_el = document.createElement('style')
      style_el.textContent = data[0]
      return style_el
    }
  }

  function ondata (data) {
    if (data.length === 0 || !data[0]) return
    button.__count = data[0].click_count || 0
    update_count()
    if (data[0].result) show_result(data[0].result)
  }

  function step_data (data) {
    button.__accessible = data.is_accessible !== false
    button.__step_index = data.index !== undefined ? data.index : 0
    button.disabled = !button.__accessible
    reset_clicks()
  }

  function reset_data () {
    reset_clicks()
    drive.put('data/form_click_rate_test.json', {
      click_count: 0,
      result: ''
    })
  }

  function on_docs_mode_change (active) {
    if (!active) update_count()
  }

  function reset_clicks () {
    button.__count = 0
    button.__start = 0
    result_el.textContent = ''
    update_count()
  }

  function update_count () { count_el.textContent = button.__count }
  function show_result (result) { result_el.textContent = result }
}

function fallback_module () {
  return {
    api: fallback_instance,
    _: {
      DOCS: { $: '' },
      net_helper: { $: '' }
    }
  }

  function fallback_instance () {
    return {
      _: {
        DOCS: { 0: '' },
        net_helper: { 0: '' }
      },
      drive: {
        'style/': {
          'form_click_rate_test.css': { $ref: 'form_click_rate_test.css' }
        },
        'data/': {
          'form_click_rate_test.json': { raw: { click_count: 0, result: '' } }
        },
        'docs/': { 'README.md': { raw: '# Click Rate Test\nClick 10 times as fast as possible.' } }
      }
    }
  }
}
