// taskbar/taskbar.js
const STATE = require('STATE')
const statedb = STATE(__filename)
const { get } = statedb(fallback_module)
const net = require('net_helper')
const action_bar = require('action_bar')
const action_executor = require('action_executor')
const tabsbar = require('tabsbar')

module.exports = taskbar

async function taskbar (opts, invite) {
  const { id, sdb } = await get(opts.sid)
  const { drive } = sdb

  const on = {
    style: inject
  }
  const on_message = {
    update_steps_wizard_for_app: handle_update_steps_wizard_for_app,
    docs_toggle: handle_docs_toggle,
    load_actions: handle_load_actions,
    step_clicked: handle_step_clicked,
    update_quick_actions_for_app: handle_update_quick_actions_for_app,
    update_quick_actions_input: handle_update_quick_actions_input,
    show_submit_btn: handle_submit_btn_toggle,
    hide_submit_btn: handle_submit_btn_toggle,
    add_link_tab: handle_forward_tabsbar,
    remove_link_tab: handle_forward_tabsbar
  }
  const { io, _ } = net(id)

  const el = document.createElement('div')
  const shadow = el.attachShadow({ mode: 'closed' })

  shadow.innerHTML = `
  <div class="taskbar-container main">
    <div class="action-executor-slot"></div>
    <div class="bottom-slot">
      <div class="action-bar-slot"></div>
      <div class="tabsbar-slot"></div>
    </div>
  </div>`
  const sheet = new CSSStyleSheet()
  shadow.adoptedStyleSheets = [sheet]
  const action_executor_slot = shadow.querySelector('.action-executor-slot')
  const action_bar_slot = shadow.querySelector('.action-bar-slot')
  const tabsbar_slot = shadow.querySelector('.tabsbar-slot')

  const subs = await sdb.watch(onbatch)
  io.on = {
    up: onmessage,
    action_bar: action_bar_protocol,
    action_executor: action_executor_protocol,
    tabsbar: tabsbar_protocol
  }
  if (invite) io.accept(invite)

  const action_bar_el = await action_bar({ ...subs[0] }, io.invite('action_bar', { up: id }))
  action_bar_el.classList.add('replaced-action-bar')
  action_bar_slot.replaceWith(action_bar_el)

  const action_executor_el = await action_executor({ ...subs[1] }, io.invite('action_executor', { up: id }))
  action_executor_el.classList.add('replaced-action-executor')
  action_executor_slot.replaceWith(action_executor_el)

  const tabsbar_el = await tabsbar({ ...subs[2] }, io.invite('tabsbar', { up: id }))
  tabsbar_el.classList.add('replaced-tabsbar')
  tabsbar_slot.replaceWith(tabsbar_el)

  return el

  async function onbatch (batch) {
    for (const { type, paths } of batch) {
      const data = await Promise.all(paths.map(load_path_raw))
      const func = on[type] || fail
      func({ data, type })
    }

    function load_path_raw (path) { return drive.get(path).then(read_drive_file_raw) }
    function read_drive_file_raw (file) { return file.raw }
  }

  function fail ({ data, type }) { console.warn('invalid message', { cause: { data, type } }) }

  function inject ({ data }) { sheet.replaceSync(data[0]) }

  // ---------
  // PROTOCOLS
  // ---------

  function action_bar_protocol (msg) {
    const action_handlers = {
      action_submitted: action_bar_forward_action_executor,
      selected_action: action_bar_forward_action_executor,
      activate_steps_wizard: action_bar_forward_action_executor,
      render_form: action_bar_forward_action_executor,
      console_history_toggle: action_bar_forward_up,
      ui_focus: action_bar_forward_up,
      display_actions: action_bar_forward_up,
      filter_actions: action_bar_forward_up
    }
    const handler = action_handlers[msg.type] || action_bar_forward_up
    handler(msg)

    function action_bar_forward_action_executor (msg) { _.action_executor(msg.type, msg.data, msg.head ? { cause: msg.head } : {}) }
    function action_bar_forward_up (msg) { _.up(msg.type, msg.data, msg.head ? { cause: msg.head } : {}) }
  }

  function action_executor_protocol (msg) {
    console.error('taskbar: action_executor_protocol', msg.type, msg.data)
    const action_handlers = {
      load_actions: action_executor__forward_action_bar,
      step_clicked: action_executor__forward_action_bar,
      show_submit_btn: action_executor__forward_action_bar,
      hide_submit_btn: action_executor__forward_action_bar,
      action_auto_completed: action_executor__auto_completed
    }
    const handler = action_handlers[msg.type] || action_executor__noop
    handler(msg)
    _.up(msg.type, msg.data, msg.head ? { cause: msg.head } : {})

    function action_executor__forward_action_bar (msg) { _.action_bar(msg.type, msg.data, msg.head ? { cause: msg.head } : {}) }
    function action_executor__noop () {}

    function action_executor__auto_completed (msg) { _.action_bar('action_submitted', msg.data, msg.head ? { cause: msg.head } : {}) }
  }

  function tabsbar_protocol (msg) {
    const action_handlers = {
      docs_toggle: tabsbar_docs_toggle,
      link_tab_close_clicked: tabsbar_forward_up,
      link_tab_clicked: tabsbar_forward_up
    }
    const handler = action_handlers[msg.type] || tabsbar__noop
    handler(msg)
    _.up(msg.type, msg.data, msg.head ? { cause: msg.head } : {})

    function tabsbar_docs_toggle (msg) {
      _.action_bar(msg.type, msg.data, msg.head ? { cause: msg.head } : {})
      _.action_executor(msg.type, msg.data, msg.head ? { cause: msg.head } : {})
    }

    function tabsbar_forward_up (msg) { _.up(msg.type, msg.data, msg.head ? { cause: msg.head } : {}) }
    function tabsbar__noop () {}
  }

  function onmessage (msg) {
    const handler = on_message[msg.type] || onmessage_forward_action_bar
    handler(msg)
  }

  function handle_update_steps_wizard_for_app (msg) { _.action_executor(msg.type, msg.data, msg.head ? { cause: msg.head } : {}) }
  function handle_docs_toggle (msg) { _.action_bar(msg.type, msg.data, msg.head ? { cause: msg.head } : {}) }
  function handle_load_actions (msg) { _.action_bar(msg.type, msg.data, msg.head ? { cause: msg.head } : {}) }
  function handle_step_clicked (msg) { _.action_bar(msg.type, msg.data, msg.head ? { cause: msg.head } : {}) }
  function handle_update_quick_actions_for_app (msg) { _.action_bar(msg.type, msg.data, msg.head ? { cause: msg.head } : {}) }
  function handle_update_quick_actions_input (msg) { _.action_bar(msg.type, msg.data, msg.head ? { cause: msg.head } : {}) }
  function handle_submit_btn_toggle (msg) { _.action_bar(msg.type, msg.data, msg.head ? { cause: msg.head } : {}) }
  function handle_forward_tabsbar (msg) { _.tabsbar(msg.type, msg.data, msg.head ? { cause: msg.head } : {}) }
  function onmessage_forward_action_bar (msg) { _.action_bar(msg.type, msg.data, msg.head ? { cause: msg.head } : {}) }
}

function fallback_module () {
  return {
    api: fallback_instance,
    _: {
      action_bar: {
        $: ''
      },
      action_executor: {
        $: ''
      },
      tabsbar: {
        $: ''
      },
      net_helper: {
        $: ''
      }
    }
  }

  function fallback_instance () {
    return {
      _: {
        action_bar: {
          0: '',
          mapping: {
            icons: 'icons',
            style: 'style',
            variables: 'variables',
            data: 'data',
            actions: 'actions',
            hardcons: 'hardcons',
            prefs: 'prefs',
            docs: 'docs'
          }
        },
        action_executor: {
          0: '',
          mapping: {
            style: 'style',
            variables: 'variables',
            docs: 'docs',
            data: 'data'
          }
        },
        tabsbar: {
          0: '',
          mapping: {
            icons: 'icons',
            style: 'style',
            docs: 'docs',
            actions: 'actions'
          }
        },
        net_helper: {
          0: ''
        }
      },
      drive: {
        'style/': {
          'theme.css': {
            raw: `
              .taskbar-container {
                display: flex;
                background: #2d2d2d;
                column-gap: 1px;
                flex-direction: column;
                align-content: center;
                justify-content: center;
                container-type: inline-size;
              }
              .replaced-tabsbar {
                display: flex;
                flex: auto;
              }
              .replaced-action-bar {
                display: flex;
                flex-direction: column;
                justify-content: center;
                align-items: flex-start;
                background: #131315;
              }
              .replaced-action-executor {
                display: flex;
              }
              .bottom-slot {
                display: flex;
                flex-direction: row;
                justify-content: space-between;
              }
              @container (max-width: 768px) {
                .bottom-slot {
                  flex-direction: column;
                }
              }
            `
          }
        },
        'icons/': {},
        'variables/': {},
        'data/': {},
        'actions/': {},
        'hardcons/': {},
        'prefs/': {},
        'docs/': {
          'README.md': {
            $ref: 'README.md'
          }
        }
      }
    }
  }
}
